sys/net/gnrc/application_layer/dns/gnrc_dns.c

```diff
@@ -0,0 +1,447 @@
+/*
+ * Copyright (C) 2015 Nick van IJzendoorn <nijzendoorn@engineering-spirit.nl>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for
+ * more details.
+ */
+
+/**
+ * @ingroup net_gnrc_dns
+ * @{
+ *
+ * @file
+ *
+ * @author      Nick van IJzendoorn <nijzendoorn@engineering-spirit.nl>
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <time.h>
+
+#include "net/gnrc/dns/internals.h"
+#include "net/gnrc/ipv6.h"
+#include "net/gnrc/netapi.h"
+#include "net/gnrc/netreg.h"
+#include "net/gnrc/pktbuf.h"
+#include "net/gnrc/udp.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+/* For PRIu16 etc. */
+#include <inttypes.h>
+#endif
+
+/**
+ * @brief   Allocate memory for the UDP thread's stack
+ */
+#if ENABLE_DEBUG
+static char _stack[GNRC_DNS_STACK_SIZE + THREAD_EXTRA_STACKSIZE_PRINTF];
+#else
+static char _stack[GNRC_DNS_STACK_SIZE];
+#endif
+
+static void *_event_loop(void *args);
+
+kernel_pid_t _dns_pid = KERNEL_PID_UNDEF;
+static gnrc_dns_request_t *_dns_requests[GNRC_DNS_MAX_SIMULTANEOUS_REQUESTS];
+static gnrc_dns_servers_t _dns; /**< DNS servers object (TODO currently only one, because we only have one interface) */
+static uint16_t _request_id;
+
+static void _gnrc_dns_delete_request(gnrc_dns_request_t *req) {
+    /* free the request from the running DNS requests */
+    for (int idx = 0; idx < sizeof(_dns_requests[0]); ++idx) {
+        if (_dns_requests[idx] == req) {
+            _dns_requests[idx] = NULL;
+            return;
+        }
+    }
+}
+
+static bool _gnrc_dns_add_request(gnrc_dns_request_t *req) {
+    /* free the request from the running DNS requests */
+    for (int idx = 0; idx < sizeof(_dns_requests[0]); ++idx) {
+        if (!_dns_requests[idx]) {
+            _dns_requests[idx] = req;
+            return true;
+        }
+    }
+
+    return false;
+}
+
+static gnrc_dns_request_t* _gnrc_dns_get_request(uint16_t id) {
+    /* free the request from the running DNS requests */
+    for (int idx = 0; idx < sizeof(_dns_requests[0]); ++idx) {
+        if (_dns_requests[idx] && _dns_requests[idx]->request_id == id) {
+            return _dns_requests[idx];
+        }
+    }
+
+    return NULL;
+}
+
+gnrc_dns_servers_t *_gnrc_dns_get_context(void) {
+    return &_dns;
+}
+
+int gnrc_dns_init(void) {
+    /* check if thread is already running */
+    if (_dns_pid == KERNEL_PID_UNDEF) {
+        /* start DNS thread */
+        _dns_pid = thread_create(_stack, sizeof(_stack), GNRC_DNS_PRIO,
+        CREATE_STACKTEST, _event_loop, NULL, "dns");
+    }
+    return _dns_pid;
+}
+
+int gnrc_dns_print(void) {
```

Comment:
You could put it to its own *.c file, then the linker will only include it, when it is used, regardless of optimization settings.